### PR TITLE
feat(core/common): Add @Search decorator to the available HTTP route handlers

### PR DESCRIPTION
**Pull request by Gustrb**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #10420

## What is the new behavior?
This PR adds the new Search method decorator to be a valid handler to routes. 
This change was rather simple, since both express and fastify already support this little akward and rather new http method

## Does this PR introduce a breaking change?
- [ ] Yes
- [X] No

## Other information

### CODE REVIEW

File: packages/core/adapters/http-adapter.ts [hunk 1]
```diff
@@ -80,6 +80,12 @@ export abstract class AbstractHttpAdapter<
     return this.instance.listen(port, hostname, callback);
   }
 
+  public search(port: string | number, callback?: () => void);
+  public search(port: string | number, hostname: string, callback?: () => void);
+  public search(port: any, hostname?: any, callback?: any) {
+    return this.instance.search(port, hostname, callback);
+  }
```
Comment (micalevisk) [inline packages/core/adapters/http-adapter.ts:87]: can we move this to right below `public all`

Comment (Gustrb) [reply packages/core/adapters/http-adapter.ts]: sure thing!

**Comment by micalevisk** [comment]
image

:thinking: 

force-push it again @Gustrb

**Comment by Gustrb** [comment]
Oops, youre right, my bad, I wanted to verify my commits.

**Comment by micalevisk** [comment]
nvm that's due to this: https://github.com/nestjs/nest/pull/10545#discussion_r1018455033

**Comment by Gustrb** [comment]
oh, makes sense

**Comment by Gustrb** [comment]
btw, should I do something to fix it or should I wait the other issue?

**Comment by micalevisk** [comment]
@Gustrb I guess you can apply the commit 9970ace66518fe5e2c740c62c808f7adf748e7a3 here as well just to triggers the pipeline

**Comment by micalevisk** [comment]
I notice that this PR introduces a breaking change as the `AbstractHttpAdapter` is public and we're addind a new public method.

---

I feel that we could find a better way to add more route decorators... Express supports others routing methods that we don't bring to the core but we could have a way to let the consumer extending it if they want to.

An ideia: having a class with factory method that will use `createMappingDecorator` to create and register your route to a global storage.

**Comment by kamilmysliwiec** [comment]
> I feel that we could find a better way to add more route decorators... Express supports others routing methods that we don't bring to the core but we could have a way to let the consumer extending it if they want to.

There's a way - you can inject the HTTP adapter and register the route manually, instead of using a dedicated decorator.

**Comment by micalevisk** [comment]
yeah... I'm thinking on how we could allow the usage of such method at controller's class as well.

@kamilmysliwiec from client POV, does having yet another decorator factory to create any http route sounds bad to you?

something like this:

```ts
const Search = HttpRoute('SEARCH')

class FooController {
   @HttpRoute('GET')('/')
   getHello() {}

   @Search()
   bar() {}
}
```

then all the native ones would be just an alias to that.

I could try to implement that while still supporting `@nestjs/swagger` plugin somehow, if you like the idea :smile_cat:

**Comment by micalevisk** [comment]
I just notice that in the past Nest had the `@RequestMapping({ method })` decorator factory. I don't know what were the reasons to remove that one but I guess we won't bring that abstraction to the core again.

**Comment by wenerme** [comment]
I don't find any reference, I think this maybe the `QUERY` method. https://www.ietf.org/archive/id/draft-ietf-httpbis-safe-method-w-body-02.html

**Pull request merged**